kernel/pipe.c

```diff
@@ -110,7 +100,7 @@ int z_impl_k_pipe_write(struct k_pipe *pipe, const uint8_t *data, size_t len, k_
 		}
 
 		if (pipe_empty(pipe)) {
-			notify_waiter(&pipe->data);
+			z_sched_wake(&pipe->data, 0, NULL);
```

Comment:
What is the difference between the `z_sched_wake` & `z_upend_*` ?

Reply:
Layering.  The former is synchronized and manages the timeout value and some other bits of state you need to worry about when you wake a thread up.